lib/fog/orchestration/util/recursive_hot_file_loader.rb

```diff
@@ -0,0 +1,243 @@
+
+          raise "template_file should be Hash or String" unless
+            template_file.kind_of?(String) || template_file.kind_of?(Hash)
+
+          local_base_url = base_url_for_url(normalise_file_path_to_url(Dir.pwd + "/TEMPLATE"))
+
+          if template_file.kind_of?(Hash)
+            template_base_url = local_base_url
+            # Serialize the template hash to deep-copy it and
+            #  avoid modifying the argument. Instead create a
+            #  new one to be modified by get_file_contents.
+            raw_template = YAML.dump(template_file)
+          elsif template_is_raw?(template_file)
+            raw_template = template_file
+            template_base_url = local_base_url
+          elsif template_is_url?(template_file)
+            template_file = normalise_file_path_to_url(template_file)
+            template_base_url = base_url_for_url(template_file)
+            raw_template = get_content(template_file)
+
+            Fog::Logger.warning("Template visited: #{@visited}")
+            @visited[template_file] = true
+          else
+            raise "template_file is not a string of the expected form"
+          end
+          template = YAML.safe_load(raw_template, [Date])
+
+          get_file_contents(template, template_base_url)
+
+          template
+        end
+
+        # Traverse the template tree looking for get_file and type
+        #   and populating the @files attribute with their content.
+        #   Resource referenced by get_file and type are eventually
+        #   replaced with their absolute URI as done in heatclient
+        #   and shade.
+        #
+        def get_file_contents(from_data, base_url = nil)
+          Fog::Logger.warning("Processing #{from_data} with base_url #{base_url}")
+
+          # Recursively traverse the tree.
+          if recurse_if(from_data)
+            recurse_data = from_data.kind_of?(Hash) ? from_data.values : from_data
+            recurse_data.each do |value|
+              get_file_contents(value, base_url)
+            end
+          end
+
+          # I'm on a Hash, process it.
+          if from_data.kind_of?(Hash)
+            from_data.each do |key, value|
+              next if ignore_if(key, value)
+              Fog::Logger.debug("Inspecting #{key}, #{value} at #{base_url}")
+
+              # Resolve relative paths.
+              base_url += '/' if base_url && !base_url.end_with?('/')
+              str_url = url_join(base_url, value)
+
+              next if @files.key?(str_url)
+              # Don't process file:// outside our base_url. TODO raise an exception here?
+              next if file_outside_base_url?(base_url, str_url)
+
+              file_content = get_content(str_url)
+
+              # get_file should not recurse hot templates.
+              if key == "type" && template_is_raw?(file_content) && !(@visited[str_url])
+                template = get_template_contents(str_url)
+                file_content = YAML.dump(template)
```

Comment:
Why do you need to convert the content to a hash and dump it back to yaml? Don't you already get the content?

Why do you say "get_file should not recurse hot templates"? What about the nested hot template contains `get_file`?

Reply:
### file_content 

If I run
```
stack create --template http://foo.com/openshift.yaml
```
and openshift.yaml contains
```
type: master.yaml
```

I expect the original template should upload to heat 
```
# modified openshift.yaml
type: http://foo.com/master.yaml
```
I'm adding a relevant test for that case.

### `type` vs `get_file`

If a template is passed as a `type' then it's part of the stack, and it's used by heat create resources.

```
resources:
  
  server:
    type: server.yaml
    ...
```

If a template is referenced via `get_file`, then it should be considered just a plain file to be attached to the templates, but is not interpreted by heat itself.

```
resources:

  config_files:
    type: OS::Heat::SoftwareConfig
    properties:
      group: script
      config:
      - get_file: uset_just_like_a_file.yaml

```